tests/foreman/cli/test_jobtemplate.py

```diff
@@ -28,183 +28,178 @@
-        self.assertGreaterEqual(len(templates), 1)
-        Defaults.add({'param-name': 'organization_id', 'param-value': self.organization['id']})
-        Defaults.add({'param-name': 'location_id', 'param-value': location['id']})
-        try:
-            templates = JobTemplate.list()
-            self.assertGreaterEqual(len(templates), 1)
-        finally:
-            Defaults.delete({'param-name': 'organization_id'})
-            Defaults.delete({'param-name': 'location_id'})
-
-    @pytest.mark.tier2
-    def test_positive_view_dump(self):
-        """Export contents of a job template
-
-        :id: 25fcfcaa-fc4c-425e-919e-330e36195c4a
-
-        :expectedresults: Verify no errors are thrown
-
-        """
-        template_name = gen_string('alpha', 7)
+    :expectedresults: The empty file is detected and error is raised
+
+    :CaseImportance: Critical
+    """
+    template_name = gen_string('alpha', 7)
+    with pytest.raises(CLIFactoryError, match='Could not create the job template:'):
         make_job_template(
             {
-                'organizations': self.organization['name'],
+                'organizations': module_org.name,
                 'name': template_name,
-                'file': TEMPLATE_FILE,
+                'file': TEMPLATE_FILE_EMPTY,
             }
         )
-        dumped_content = JobTemplate.dump({'name': template_name})
-        self.assertGreater(len(dumped_content), 0)
+
+
+@pytest.mark.tier1
+@pytest.mark.upgrade
+def test_positive_delete_job_template(module_org):
+    """Delete a job template
+
+    :id: 33104c04-20e9-47aa-99da-4bf3414ea31a
+
+    :expectedresults: The Job Template has been deleted
+
+    :CaseImportance: Critical
+    """
+    template_name = gen_string('alpha', 7)
+    make_job_template(
+        {
+            'organizations': module_org.name,
+            'name': template_name,
+            'file': TEMPLATE_FILE,
+        }
+    )
+    JobTemplate.delete({'name': template_name})
+    with pytest.raises(CLIReturnCodeError):
+        JobTemplate.info({'name': template_name})
+
+
+@pytest.mark.run_in_one_thread
+@pytest.mark.tier2
+def test_positive_list_job_template_with_saved_org_and_loc(module_org):
+    """List available job templates with saved default organization and
+    location in config
+
+    :id: 4fd05dd7-53e3-41ba-ba90-6181a7190ad8
+
+    :expectedresults: The Job Template can be listed without errors
+
+    :BZ: 1368173
+    """
+    template_name = gen_string('alpha')
+    location = make_location()
+    make_job_template(
+        {
+            'organizations': module_org.name,
+            'name': template_name,
+            'file': TEMPLATE_FILE,
+        }
+    )
+    templates = JobTemplate.list({'organization-id': module_org.id})
+    assert len(templates) >= 1
+    Defaults.add({'param-name': 'organization_id', 'param-value': module_org.id})
+    Defaults.add({'param-name': 'location_id', 'param-value': location['id']})
```

Comment:
I think the try block should be from the Default add, in case if the default add get fails on location_id then the finally won't be run

Reply:
@omkarkhatavkar you raise a good point about the test flow on failure - will the call to `Defaults.delete` potentially raise an exception themselves.

I think this is a bit of an abused case of try/finally, when the pytest best practice for the intended behavior is the use of `request.addfinalizer` in the form of a function call or decoration.

If there is an exception in the call to `Defaults.add`, the test will result in an error, but that also means the finally block will never run anyway.

If there is a silent failure mode in calling `Defaults.add`, the test will continue, and hit an exception either within the try block or when `Defaults.delete` is called.

I think the best course of action here would be to follow a pattern as outlined below. The teardown phase of the pytest test loop will be responsible for handling the delete of the `Default` entities, and will be safe from any silent failures.  Its a pattern that also is applicable to test functions where entities are sometimes deleted during the course of the test (parametrization, general use fixtures creating entities).

```python
# test function signature updated to add request
# job templates listing, test content up to L176
Defaults.add({'param-name': 'organization_id', 'param-value': module_org.id})
Defaults.add({'param-name': 'location_id', 'param-value': location['id']})

@request.addfinalizer
def defaults_teardown():
    try:
        Defaults.delete({'param-name': 'organization_id'})
    except:
        # maybe log that teardown failed here, since the fixture scope in the framework is unique to this test
        pass
    try:
        Defaults.delete({'param-name': 'location_id'})
    except:
        pass

# Now get on with test content, the above function will execute on teardown
templates = JobTemplate.list()
assert len(templates) >= 1 

```

Reply:
@pondrejk make sense if that unwanted test. I gave you a suggestion on the code. If it is not a valid test step then you should be good!